File: gremlin-java/src/test/java/com/tinkerpop/gremlin/java/FunctionStepTest.java
package com.tinkerpop.gremlin.java;

import com.tinkerpop.blueprints.Graph;
import com.tinkerpop.blueprints.Vertex;
import com.tinkerpop.blueprints.impls.tg.TinkerGraphFactory;
import com.tinkerpop.gremlin.test.ComplianceTest;
import com.tinkerpop.pipes.PipeFunction;

import java.util.Iterator;

/**
 * @author Marko A. Rodriguez (http://markorodriguez.com)
 */
public class FunctionStepTest extends com.tinkerpop.gremlin.test.FunctionStepTest {

    Graph g = TinkerGraphFactory.createTinkerGraph();

    public void testCompliance() {
        ComplianceTest.testCompliance(this.getClass());
    }

    public void test_g_v1_out_stepXnext_nameX() {
        super.test_g_v1_out_stepXnext_nameX(new GremlinPipeline(g.getVertex(1)).out().step(new PipeFunction<Iterator<Vertex>, Vertex>() {
            public Vertex compute(Iterator<Vertex> iterator) {
                return iterator.next();
            }
        }).property("name"));

        super.test_g_v1_out_stepXnext_nameX(new GremlinPipeline(g.getVertex(1)).optimize(false).out().step(new PipeFunction<Iterator<Vertex>, Vertex>() {
            public Vertex compute(Iterator<Vertex> iterator) {
                return iterator.next();
            }
        }).property("name"));
    }
}

File: gremlin-scala/src/test/scala/com/tinkerpop/gremlin/scala/FunctionStepTest.scala
package com.tinkerpop.gremlin.scala

import com.tinkerpop.blueprints.pgm.Vertex
import com.tinkerpop.blueprints.pgm.impls.tg.TinkerGraphFactory
import com.tinkerpop.gremlin.test.ComplianceTest
import java.util.{Iterator => JIterator}

class FunctionStepTest extends com.tinkerpop.gremlin.test.FunctionStepTest {
  val g = TinkerGraphFactory.createTinkerGraph()

  override def testCompliance() {
    ComplianceTest.testCompliance(this.getClass)
  }

  def test_g_v1_out_stepXnext_nameX() {
    //TODO can we use a scala Iterator here instead of JIterator?
    super.test_g_v1_out_stepXnext_nameX(g.v(1).out.step({it: JIterator[Vertex] => it.next()("name").toString}))
  }
}
